Add support for LPCXpresso54114

MaureenHelm opened the pull request:
Rebased #1101 and converted to CMake. ~~This is not yet ready to merge,~~ but making it available since there is an increasing interest in enabling [RPMsg-Lite](https://github.com/NXPmicro/rpmsg-lite) and [eRPC](https://github.com/EmbeddedRPC/erpc) in Zephyr.

There are still a few outstanding issues to resolve:
- [x] USART drops the last few characters in every printk
- [x] Clean up naming conventions to be consistent with other mcux drivers
- [x] Clean up formatting

We will also need to figure out how to enable the second core, but that will likely be in another PR.

Commit:
lpc: Add device header and driver files for lpc54114 from mcux 2.2.1

Renamed fsl drivers for lpc board as fsl_lpc drivers.

Origin: NXP MCUXpresso SDK 2.2.1
URL: mcux.nxp.com
Maintained-by: External

Signed-off-by: Shiksha Patel <shiksha.patel@nxp.com>
Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

Commit:
lpc: Add nxp_lpc soc family to soc directory

Add soc configuration support for lpc soc family, Kconfigs and soc files
for lpcxxx soc.

Add dtsi file for lpc54xxx.

Signed-off-by: Shiksha Patel <shiksha.patel@nxp.com>
Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

Commit:
lpc: Add pinmux shim driver for lpcxpresso54114

Adds a new pinmux driver for lpcxpresso54114 based on mcux.

Signed-off-by: Shiksha Patel <shiksha.patel@nxp.com>
Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

Commit:
lpc: Add usart shim driver for lpcxpresso54114

Add usart-yaml in dts.

Build fsl_lpc_usart and fsl_lpc_flexcomm in
ext/hal/nxp/mcux/drivers/Makefile.

Only polling mode is implemented in usart now. Interrupt can be added in
future.

Signed-off-by: Shiksha Patel <shiksha.patel@nxp.com>
Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

Commit:
lpc: Add gpio mcux driver for lpc

Add gpio mcux driver which can be used for lpcxpresso54114 and other lpc
socs. In this driver, CMSIS register access is made for GPIO.

Option for access by GPIO Pin is provided as of now.

Signed-off-by: Shiksha Patel <shiksha.patel@nxp.com>
Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

Commit:
lpc: Add lpcxpresso54114 board based on its architecture

Add default drivers, pinmux configurations, and other board level
configuration parameters in respective Kconfigs.

Add board dts, fixup, and docs.

Samples tested include:
- hello_world
- sychronization
- basic/blinky
- basic/threads

Signed-off-by: Shiksha Patel <shiksha.patel@nxp.com>
Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm commented:
Cleaned up naming and formatting. I think I've also resolved the USART issue by updating the LPC-Link firmware, but want to validate this on another board.

dbkinder reviewed (dismissed):
doc LGTM

MaureenHelm commented:
Rebased and fixed build error caused by dts/yaml changes in 104553d

carlescufi commented:
@skordal FYI

MaureenHelm commented:
@galak, this is ready for review. Please have a look

galak reviewed (commented):
In general, wondering if we should have copyright 2017-2018 on this

galak commented on arch/arm/soc/nxp_lpc/lpc54xxx/Kconfig.defconfig.lpc54114:
```diff
@@ -0,0 +1,38 @@
+config GPIO
+	def_bool y
+
+if GPIO
```
since you always enable GPIO, is the if needed? (when the 'if GPIO' goes aways this should probably also be a select in Kconfig.series

galak commented on arch/arm/soc/nxp_lpc/lpc54xxx/Kconfig.defconfig.lpc54114:
```diff
@@ -0,0 +1,38 @@
+
+endif # PINMUX
+
+config GPIO
```
Should maybe just convert this to a select in Kconfig.series

MaureenHelm replied:
No, there may be an application that does not need gpio and wants to turn it off.

galak replied:
How would the app turn off GPIO?

MaureenHelm replied:
set `CONFIG_GPIO=n` in proj.conf

galak commented on arch/arm/soc/nxp_lpc/lpc54xxx/soc.h:
```diff
@@ -0,0 +1,35 @@
+}
+#endif
+
+#define BOARD_BOOTCLOCKFROHF48M_CORE_CLOCK		48000000U
```
are we not able to get this from clock_config.h?

MaureenHelm replied:
We can remove it and use CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC in soc.c instead

galak commented on boards/arm/lpcxpresso54114/Kconfig.defconfig:
```diff
@@ -0,0 +1,43 @@
+
+endif # USART_MCUX_LPC
+
+config PINMUX
```
make this a select in Kconfig.board

MaureenHelm replied:
Making it a select causes warnings:
```
warning: (BOARD_LPCXPRESSO54114) selects PINMUX which has unmet direct dependencies (BOARD_FRDM_K64F || BOARD_FRDM_KL25Z || BOARD_FRDM_KW41Z || BOARD_HEXIWEAR_K64 || BOARD_USB_KW24D512 || BOARD_GALILEO)
```

Adding `CONFIG_PINMUX=y` to lpcxpresso54114_defconfig eliminates the warnings and is more consistent with other boards.

galak commented on boards/arm/lpcxpresso54114/lpcxpresso54114.dts:
```diff
@@ -0,0 +1,29 @@
+#include <nxp/nxp_lpc54xxx.dtsi>
+
+/ {
+	model = "NXP LPCXPRESSO54114 board";
```
nit: model = "NXP LPCXpresso54114 board";

galak commented on boards/arm/lpcxpresso54114/lpcxpresso54114.dts:
```diff
@@ -0,0 +1,29 @@
+
+/ {
+	model = "NXP LPCXPRESSO54114 board";
+	compatible = "nxp,lpc54xxx","nxp,lpc";
```
nit, space after ','

galak commented on boards/arm/lpcxpresso54114/lpcxpresso54114.dts:
```diff
@@ -0,0 +1,29 @@
+
+	aliases{
+		usart_0 = &usart0;
+		};
```
nit: '};' should have one less tab.

galak commented on boards/arm/lpcxpresso54114/lpcxpresso54114.dts:
```diff
@@ -0,0 +1,29 @@
+	aliases{
+		usart_0 = &usart0;
+		};
+	chosen {
```
nit: blank line after };

galak commented on boards/arm/lpcxpresso54114/lpcxpresso54114.yaml:
```diff
@@ -0,0 +1,15 @@
+flash: 256
+toolchain:
+  - zephyr
+  - gccarmemb
```
Should probably add a 'supported' section w/'gpio'

galak reviewed (changes requested):
Minor things w/kconfig and spacing.

MaureenHelm commented:
> In general, wondering if we should have copyright 2017-2018 on this

I consulted with an internal adviser and they said it wasn't necessary to append 2018 in this case.

MaureenHelm commented:
- Rebased to latest arm branch and resolved some new jlink runner warnings caused by introduction of e33ec242
- Removed `BOARD_BOOTCLOCKFROHF48M_CORE_CLOCK` from soc.h and replaced references with `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC`
- Moved pinmux default from Kconfig.defconfig.lpc54114 to lpcxpresso54114_defconfig
- Fixed dts code alignment

galak commented:
Have you tried the board with CMSIS-DAP/pyOCD?

MaureenHelm commented:
> Have you tried the board with CMSIS-DAP/pyOCD?

Not yet. pyOCD didn't support this SoC when Shiksha originally wrote these patches. I'd prefer to add in a subsequent patch to avoid making more than cosmetic changes to her patch series.

galak commented:
Are you using jlink to flash?  Can/should we expand the jlink runner to support flashing?

carlescufi commented:
@galak J-Link already supports flashing

galak commented on arch/arm/soc/nxp_lpc/lpc54xxx/soc.h:
```diff
@@ -0,0 +1,33 @@
+
+#ifndef _ASMLANGUAGE
+#include <device.h>
+#include <misc/util.h>
```
do we need misc/util.h included twice?

MaureenHelm replied:
Probably not. :)

galak commented on boards/arm/lpcxpresso54114/doc/lpcxpresso54114.rst:
```diff
@@ -0,0 +1,168 @@
+Before you start using Zephyr on the LPCXpresso54114, download and run
+`LPCScrypt`_ to update the LPC-Link2 firmware to the latest version. Serial
+communication problems, such as dropping characters, have been observed with
+older versions of the firmware.
```
can we spec a version here about the firmware.  I'm wondering when someone reads this a year from now or two, will the comment still be valid?  Seems like if we can reference some versions it would be helpful.  (ie needing firmware version > X.Y.Z)

MaureenHelm replied:
Problem is that I don't know what firmware version was on the board before I reflashed it. I can specify a version that I know works, though that may be a newer version than really required.

galak replied:
I think at least doing that would be good.  Saying something about older version might be ok, but good know version is X.Y.Z.

galak reviewed (changes requested):
Some minor things.  Would be good to have at least one supported flashing solution documented.

galak commented:
> EDIT: I might be wrong here, nrfjprog, which uses J-Link under the hood does support flashing, but I'm not sure about raw J-Link. @mbolivar?

not sure, @d3zd3z gave me this info on how he's using j-link:
https://gist.github.com/d3zd3z/1b104bad90526df49b7a915248ff82f8

MaureenHelm commented:
> Are you using jlink to flash? Can/should we expand the jlink runner to support flashing?

Yes I'm using jlink to flash, but via `make debug`. Yes, we should expand the jlink runner to support `make flash`. I started to look at that right before the holidays and had a prototype working for LPC54114, but it didn't work with Kinetis.

mbolivar commented:
@carlescufi: what @MaureenHelm said.

I'd be happy to see flash supported in jlink.py, especially with `flash_addr=True` in the `RunnerCaps` (for making life with MCUBoot easier).

MaureenHelm commented:
> I'd be happy to see flash supported in jlink.py, especially with flash_addr=True in the RunnerCaps (for making life with MCUBoot easier).

Created new issue #5755

mbolivar commented:
> Created new issue #5755

Ah, thanks for that

MaureenHelm commented:
- Rebased to latest arm branch
- Removed duplicate `#include <misc/util.h>` in soc.h
- Updated board doc with segger firmware version
